### ISSUE TITLE
register-spill regression in the gradient of einsum with ellipsis in TF 2.21  vs 2.20

### ISSUE THREAD

**Issue by AntBoc** [tensorflow/tensorflow#122274]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

binary

### TensorFlow version

v2.21.0-rc1-5-ga481b10260d 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 20.04.5 / Linux Mint 22

### Python version

3.12.13

### CUDA/cuDNN version

Driver: 13.2.0; Runtime: 12.9.0; Toolkit: 12.5.0; DNN: 9.20.0

### Current behavior?

A plain dense layer y = x @ w can be written as either
    tf.einsum("...k,...kn->...n", x, w)   or
    tf.matmul(x, w),
which are equivalent in principle.
For a large batch the (double) backwards of both forms works fine in TF 2.20 producing only tiny register spills (see below).
In TF 2.21 einsum produces large spills (much larger ~200+ KB in a real workload). 
With tf.matmul spilling is largely reduced, but not uniformly across GPU architectures.
 
Observed (across A100, L40S, H200 GPUs), max "bytes spill stores":
    TF 2.20.0            einsum:     8-84 B     matmul:  44-864 B  
    TF 2.21.0            einsum: 48-71 KB     matmul:  0-12 KB*
    TF 2.22.0-dev     einsum: 52-89 KB     matmul:  0.1-12 KB*
 
* Additionally setting  --xla_gpu_enable_triton_gemm=false removes spills entirely. Small effect on tf.einsum

### Standalone code to reproduce the issue

```shell
import os

os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "0")
import tensorflow as tf

MODE = os.environ.get("MODE", "einsum")
N, K, M = 130_000, 256, 128          # N = batch (large), K = in features, M = out features
x = tf.random.normal([N, K], dtype=tf.float32)
w = tf.Variable(tf.random.normal([K, M], dtype=tf.float32))


@tf.function(jit_compile=True)
def step(x):
    # second-order produces more spills, but the same story as a
    # single backward.
    with tf.GradientTape() as t2:
        with tf.GradientTape() as t1:
            t1.watch(x)
            if MODE == "matmul":
                y = tf.matmul(x, w)
            else:
                y = tf.einsum("...k,...kn->...n", x, w)
            e = tf.reduce_sum(y * y)
        f = t1.gradient(e, x)
        loss = tf.reduce_sum(f * f)
    return t2.gradient(loss, w)


print(f"TF {tf.__version__}  MODE={MODE}", flush=True)
g = step(x)
print("ok, |dw| =", float(tf.reduce_sum(tf.abs(g))), flush=True)
```

### Relevant log output

```shell
TF 2.20.0  MODE=einsum

2026-06-29 14:08:10.852462: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_78', 4 bytes spill stores, 4 bytes spill loads
2026-06-29 14:08:10.872671: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_65', 8 bytes spill stores, 8 bytes spill loads
2026-06-29 14:08:10.922687: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_70', 40 bytes spill stores, 40 bytes spill loads
2026-06-29 14:08:11.095687: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_78', 76 bytes spill stores, 76 bytes spill loads


TF 2.20.0  MODE=matmul

2026-06-29 14:23:56.206755: I external/local_xla/xla/stream_executor/cuda/subprocess_compilation.cc:346] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_41', 864 bytes spill stores, 864 bytes spill loads


##########################################


TF 2.21.0  MODE=einsum

I0000 00:00:1782735899.496275 3968827 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_Einsum_10', 11148 bytes spill stores, 11220 bytes spill loads
I0000 00:00:1782735900.293409 3968835 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_Einsum_6', 10292 bytes spill stores, 10340 bytes spill loads
I0000 00:00:1782735901.156542 3968825 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_Einsum_8', 9656 bytes spill stores, 9844 bytes spill loads
E0000 00:00:1782736021.185085 3968838 slow_operation_alarm.cc:73] 
********************************
[Compiling module input_reduce_fusion for GPU] Very slow compile? If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
I0000 00:00:1782736025.963618 3968837 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'input_reduce_fusion', 52568 bytes spill stores, 71000 bytes spill loads
E0000 00:00:1782736025.974365 3968837 slow_operation_alarm.cc:140] The operation took 2m4.789475133s
********************************
[Compiling module input_reduce_fusion for GPU] Very slow compile? If you want to file a bug, run with envvar XLA_FLAGS=--xla_dump_to=/tmp/foo and attach the results.
********************************
I0000 00:00:1782736142.172976 3968837 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'input_reduce_fusion_1', 52380 bytes spill stores, 70668 bytes spill loads


TF 2.21.0  MODE=matmul

I0000 00:00:1782736698.649643 3969114 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_MatMul_10', 12296 bytes spill stores, 12600 bytes spill loads
I0000 00:00:1782736699.729714 3969118 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_MatMul_1_4', 64 bytes spill stores, 64 bytes spill loads
I0000 00:00:1782736699.815458 3969125 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_MatMul_1_4', 612 bytes spill stores, 436 bytes spill loads
I0000 00:00:1782736700.620576 3969121 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_MatMul_6', 12296 bytes spill stores, 12600 bytes spill loads
I0000 00:00:1782736701.402268 3969112 subprocess_compilation.cc:348] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_MatMul_8', 11384 bytes spill stores, 11512 bytes spill loads
```

**Comment by yash4515** [comment]
.take

**Comment by yash4515** [comment]
hi @Venkat6871 are you working on this issue if not please do assign to me ..i want to contribute in this
